tools/preproc/preproc.cpp

```diff
@@ -123,57 +140,65 @@ char* GetFileExtension(char* filename)
     return extension;
 }
 
+static void UsageAndExit(const char *program)
+{
+    std::fprintf(stderr, "Usage: %s [-i] [-e] SRC_FILE CHARMAP_FILE (optional)\nwhere -i denotes if input is from stdin\n      -e enables enum handling\n", program);
+    std::exit(EXIT_FAILURE);
+}
+
 int main(int argc, char **argv)
 {
-    if (argc < 2 || argc > 4)
+    int opt;
+    const char *source = NULL;
+    bool isStdin = false;
+    bool doEnum = false;
+
+    /* preproc [-i] [-e] SRC_FILE CHARMAP_FILE */
+    while ((opt = getopt(argc, argv, "ie")) != -1)
     {
-        std::fprintf(stderr, "Usage: %s SRC_FILE CHARMAP_FILE(optional) [-i]\nwhere -i denotes if input is from stdin\n", argv[0]);
-        return 1;
+        switch (opt)
+        {
+        case 'i':
+            isStdin = true;
+            break;
+        case 'e':
+            doEnum = true;
+            break;
+        default:
+            UsageAndExit(argv[0]);
+            break;
+        }
     }
-
-    if (argc == 2
-        || (argv[2][0] == '-' && argv[2][1] == 'i' && argv[2][2] == '\0')) // ./preproc xxx.c or ./preproc xxx.c -i
+    if (optind + 1 == argc) {
         g_charmap = new (std::nothrow) Charmap("");
-    else
-        g_charmap = new (std::nothrow) Charmap(argv[2]);
-    if (!g_charmap)
-        FATAL_ERROR("Failed to allocate space for Charmap.\n");
```

Comment:
I think we can still keep the null check

Reply:
Done!